File: framework/dotnet/BetterPortal/Feeds.cs
using AnyVali;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal;

/// <summary>A subscriber transport of an existing GET operation, with typed input/event contracts.</summary>
public abstract class SseFeed
{
    public Handler Owner { get; }
    public string ViewId { get; }
    public Schema InputSchema { get; }
    public Schema EventSchema { get; }
    public IReadOnlyList<Renderer> Renderers { get; }
    protected SseFeed(Handler owner, string viewId, Schema input, Schema @event, IEnumerable<Renderer> renderers)
    {
        Owner = owner; ViewId = viewId; InputSchema = input; EventSchema = @event;
        Renderers = Renderer.Unique(renderers);
        foreach (var renderer in Renderers)
        {
            if (renderer.Identity.Kind != "fragment" || renderer.Identity.Status != 200)
                throw new ArgumentException("SSE event renderers require a success fragment");
            if (!owner.Renderers.Any(value => value.Identity == renderer.Identity))
                throw new ArgumentException("An SSE event renderer requires its owning GET fragment");
        }
    }
    public abstract RawResponse OpenStream(RequestContext context, IReadOnlyDictionary<string, object?> values, string? fragment,
        Func<string, object?, object?, RenderContext> renderContext, CancellationToken cancellation = default);
}

public sealed class SseFeed<TInput, TEvent> : SseFeed
{
    private readonly Func<EventScope, TInput, CancellationToken, ValueTask<TInput>> publish;
    private readonly Func<RequestContext, IReadOnlyDictionary<string, object?>, string?, Func<string, object?, object?, RenderContext>, CancellationToken, RawResponse> open;
    private SseFeed(Handler owner, string viewId, Schema input, Schema @event, IEnumerable<Renderer> renderers,
        Func<EventScope, TInput, CancellationToken, ValueTask<TInput>> publish,
        Func<RequestContext, IReadOnlyDictionary<string, object?>, string?, Func<string, object?, object?, RenderContext>, CancellationToken, RawResponse> open)
        : base(owner, viewId, input, @event, renderers) { this.publish = publish; this.open = open; }

    public static SseFeed<TInput, TEvent> Bind<TParams, TQuery, THeaders, TBody>(Handler<TParams, TQuery, THeaders, TBody> owner,
        SseRoute<TInput, TEvent, HandlerContext<TParams, TQuery, THeaders, TBody>> route, IEnumerable<Renderer<TEvent>>? renderers = null)
    {
        var entries = (renderers ?? []).ToArray();
        return new(owner, route.ViewId, route.InputSchema, route.EventSchema, entries, route.Publish, (context, values, fragment, renderContext, cancellation) =>
        {
            var prepared = owner.Prepare(context, values, cancellation);
            Func<TEvent, CancellationToken, ValueTask<string>>? render = null;
            if (fragment is not null)
            {
                var theme = (context.Scope.App.GetValueOrDefault("shell") as Node)?.GetValueOrDefault("renderer") as string;
                var renderer = Renderer.Select(entries, theme, "fragment", fragment);
                var presentation = renderContext(renderer.Identity.Renderer, prepared.Params, prepared.Query);
                render = (value, signal) => renderer.Render(value, presentation);
            }
            return new RawResponse(route.Wire(new(context.Scope.TenantId, context.Scope.AppId), prepared, render, cancellation: cancellation),
                headers: new Dictionary<string, string> { ["content-type"] = "text/event-stream; charset=utf-8", ["cache-control"] = "no-store" });
        });
    }
    public override RawResponse OpenStream(RequestContext context, IReadOnlyDictionary<string, object?> values, string? fragment,
        Func<string, object?, object?, RenderContext> renderContext, CancellationToken cancellation = default) => open(context, values, fragment, renderContext, cancellation);
    public ValueTask<TInput> Publish(EventScope scope, TInput value, CancellationToken cancellation = default) => publish(scope, value, cancellation);
}

File: framework/python/betterportal/feeds.py
"""Bind scoped subscriber delivery to an existing typed GET handler."""
from __future__ import annotations

from typing import Any, Awaitable, Callable, Generic, Iterable, Mapping

from .handler import Params, Query, Headers, Body, Handler, HandlerContext, RequestContext
from .rendering import Renderer, RenderContext, renderers as unique_renderers, select
from .response import RawHandler, RawResponse
from .sse import Input, Event, SseRoute, EventScope


class SseFeed(Generic[Params, Query, Headers, Body, Input, Event]):
    def __init__(self, owner: Handler[Params, Query, Headers, Body, Any] | RawHandler[Params, Query, Headers, Body],
                 route: SseRoute[Input, Event, HandlerContext[Params, Query, Headers, Body]], *, renderers: Iterable[Renderer[Event]] = ()):
        self.owner, self.route = owner, route
        self.renderers = unique_renderers(renderers)
        for renderer in self.renderers:
            if renderer.identity[1] != "fragment" or renderer.identity[3] != 200:
                raise ValueError("SSE event renderers require a success fragment")
            if not any(value.identity == renderer.identity for value in owner.renderers):
                raise ValueError("An SSE event renderer requires its owning GET fragment")

    def open_stream(self, context: RequestContext, values: Mapping[str, Any], *, fragment: str | None = None,
                    render_context: Callable[[str, Any, Any], RenderContext]) -> RawResponse:
        prepared = self.owner.prepare(context, values)
        render: Callable[[Event], Awaitable[str]] | None = None
        if fragment is not None:
            theme = context.scope.app.get("shell", {}).get("renderer")
            renderer = select(self.renderers, theme, "fragment", fragment)
            presentation = render_context(renderer.identity[0], prepared.params, prepared.query)
            render = lambda value: renderer.render(value, presentation)
        return RawResponse(self.route.wire(EventScope(context.scope.tenant_id, context.scope.app_id), prepared, render=render),
            headers={"content-type": "text/event-stream; charset=utf-8", "cache-control": "no-store"})

    async def publish(self, scope: EventScope, value: Input) -> Input:
        """Service code supplies the trusted publication scope; no event history is implied."""
        return await self.route.publish(scope, value)
